core/cont/src/TList.cxx

```diff
@@ -696,49 +699,47 @@ TObjLink *TList::NewOptLink(TObject *obj, Option_t *opt, TObjLink *prev)
 
 void TList::RecursiveRemove(TObject *obj)
 {
-   R__COLLECTION_WRITE_GUARD();
-
    if (!obj) return;
 
-   // When fCache is set and has no previous and next node, it represents
-   // the node being cleared and/or deleted.
-   if (fCache && fCache->fNext == 0 && fCache->fPrev == 0) {
-      TObject *ob = fCache->GetObject();
-      if (ob && ob->TestBit(kNotDeleted)) {
+   // When fCache is set and has no previous and next node,
+   // it represents the node being cleared and/or deleted.
+   if (fCache && fCache->fNext == 0 && fCache->fPrev == 0)
+      if (TObject *ob = fCache->GetObject())
+         if (ob->TestBit(kNotDeleted))
+            ob->RecursiveRemove(obj);
+
+   for (TObjLink* lnk = fFirst; lnk; lnk = lnk->Next()) {
+      TObject *ob = lnk->GetObject();
+
+      // skip if deleted or being deleted
+      if (!ob || !ob->TestBit(kNotDeleted)) continue;
+
+      // not this object, but may contain this object
+      if (!ob->IsEqual(obj)) {
          ob->RecursiveRemove(obj);
+         continue;
       }
-   }
 
-   TObjLink *lnk  = fFirst;
-   TObjLink *next = 0;
-   while (lnk) {
-      next = lnk->Next();
-      TObject *ob = lnk->GetObject();
-      if (ob->TestBit(kNotDeleted)) {
-         if (ob->IsEqual(obj)) {
-            if (lnk == fFirst) {
-               fFirst = next;
-               if (lnk == fLast)
-                  fLast = fFirst;
-               else
-                  fFirst->fPrev = 0;
-               DeleteLink(lnk);
-            } else if (lnk == fLast) {
-               fLast = lnk->Prev();
-               fLast->fNext = 0;
-               DeleteLink(lnk);
-            } else {
-               lnk->Prev()->fNext = next;
-               lnk->Next()->fPrev = lnk->Prev();
-               DeleteLink(lnk);
-            }
-            fSize--;
-            fCache = 0;
-            Changed();
-         } else
-            ob->RecursiveRemove(obj);
+      // object found, remove it from the list
+      R__COLLECTION_WRITE_GUARD();
```

Comment:
Same here, if in the future this becomes a real lock, it's better to only really lock when actually changing the structure than during the entire traversal of it, otherwise there will be a lot of contention for the list's first element. If performance isn't great now, it will become terrible later...

Reply:
Ah.  It will not become a lock in the future.  The semantic is different from that of a lock.  It is a collision detection mechanism, you do want to have it 'held' by the function as long as possible to increase the likelyhood of overlaps when there is one.

Reply:
Alright, I will put them back where they were then. However, I'd say it's better to have lower performance than crashes... We ought to make things thread-safe in general. If it's just for lists, it shouldn't be too hard to do it while [keeping most of the performance](https://youtu.be/1Q-RH2tiyt0?t=25m32s).

Reply:
But then we should also re-implement `vector` to be thread safe. And `TObject` - all its non-`const` functions must be callable from different threads. And `int` itself should really become thread-safe, after all "it's better to have lower performance than crashes..."! Or maybe not? ;-)

ROOT's collections, just like STL collections, are non-thread-safe. That by itself is not a problem, as long as users know what they get. These collections have a different role than the role you want to assign to them.

Reply:
You really love distorting what people say. If a user runs a simple script with ROOT with MT enabled and it crashes on the list of cleanups, whose fault is it? The user is not using TList, yet ROOT crashes. We have to fix that. It's not acceptable.

Reply:
If users use TList, etc, and shoot themselves on the foot, that's obviously another story. But ROOT cannot shoot their foot for them.

Reply:
>  We have to fix that.

Yes. Indeed and for the list of cleanups we know what we are up against.

The question those guard are intending on answering is "is there *any* TList anywhere in the ROOT source code that is inadvertently/unexpectedly used from multiple thread".

I.e. after we solve the list of cleanups problems and we still get random crash, this tool will let you know whether this is because of a TList or not ... :)

Reply:
The approach so far was mostly to duplicate these collections per thread, i.e. have thread-local objects (e.g. `gDirectory`).

We have to fix crashes in the cleanup, absolutely - I just hope we can find a cheaper way than "eagerly" locking / RCUing all collections.

Reply:
Now we are talking the same language :-) This is totally about the list of cleanups, which happens to be using TList and maybe also THashList. We wil either have to move away from those containers, or use them with some sort of synchronization mechanism. I'm not aware of any other way to solve the problem. Since Philippe mentioned adding a read/write lock, I thought the solution would be the latter. However, my preferred solution is to wrap the removals into a function that delegates deletion to somebody else without blocking. For example, you could have `ROOT::DestroyObject()` which puts the object's ID on a queue of transactions that is processed by a garbage collection thread, which is woken up once there's enough stuff to be destroyed.

Reply:
> Since Philippe mentioned adding a read/write lock, I thought the solution would be the latter. 

Yes, I will be adding locks ... to new classes (THashListIWithRWLocks).

> However, my preferred solution ...

This solution has some merit and some thing to be discuss (on whether it accomplish the goals/requirement or not) ... but even before that, it is a challenging option to consider as it would require user code change ... and if the user is changing their code, I would rather have them move to v7.